### PR TITLE
logging: adds FTP logging endpoint

### PR DESCRIPTION
**Pull request by mccurdyc**
## Relevant (Referenced) Link(s)

* [Fastly API Docs](https://docs.fastly.com/api/logging#api-section-logging_ftp)
* [fastly/go-fastly](https://github.com/fastly/go-fastly/blob/c2c2c62210800daab641161412fbe210004aac0a/fastly/ftp.go)

## Validation

```
$ make test
```

```
git clone git@github.com:mccurdyc/cli.git /tmp/cli && \
( 
    cd /tmp/cli && \
    git checkout mccurdyc/logging-ftp && \
    make test && \
    rm -rf /tmp/cli \
)
```

## TODOs

- [x] Wait for next go-fastly release to include https://github.com/fastly/go-fastly/pull/165

### CODE REVIEW

File: pkg/logging/ftp/create.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	manifest manifest.Data
+	Input    fastly.CreateFTPInput
+}
+
+// NewCreateCommand returns a usable command registered under the parent.
+func NewCreateCommand(parent common.Registerer, globals *config.Data) *CreateCommand {
+	var c CreateCommand
+	c.Globals = globals
+	c.manifest.File.Read(manifest.Filename)
+	c.CmdClause = parent.Command("create", "Create an  FTP logging endpoint on a Fastly service version").Alias("add")
```
Comment (phamann) [inline pkg/logging/ftp/create.go:26]:
```suggestion
	c.CmdClause = parent.Command("create", "Create an FTP logging endpoint on a Fastly service version").Alias("add")
```

File: pkg/logging/ftp/delete.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	manifest manifest.Data
+	Input    fastly.DeleteFTPInput
+}
+
+// NewDeleteCommand returns a usable command registered under the parent.
+func NewDeleteCommand(parent common.Registerer, globals *config.Data) *DeleteCommand {
+	var c DeleteCommand
+	c.Globals = globals
+	c.manifest.File.Read(manifest.Filename)
+	c.CmdClause = parent.Command("delete", "Delete a FTP logging endpoint on a Fastly service version").Alias("remove")
```
Comment (phamann) [inline pkg/logging/ftp/delete.go:26]:
```suggestion
	c.CmdClause = parent.Command("delete", "Delete an FTP logging endpoint on a Fastly service version").Alias("remove")
```

File: pkg/logging/ftp/describe.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	manifest manifest.Data
+	Input    fastly.GetFTPInput
+}
+
+// NewDescribeCommand returns a usable command registered under the parent.
+func NewDescribeCommand(parent common.Registerer, globals *config.Data) *DescribeCommand {
+	var c DescribeCommand
+	c.Globals = globals
+	c.manifest.File.Read(manifest.Filename)
+	c.CmdClause = parent.Command("describe", "Show detailed information about a FTP logging endpoint on a Fastly service version").Alias("get")
```
Comment (phamann) [inline pkg/logging/ftp/describe.go:26]:
```suggestion
	c.CmdClause = parent.Command("describe", "Show detailed information about an FTP logging endpoint on a Fastly service version").Alias("get")
```

File: pkg/logging/ftp/update.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	Placement         common.OptionalString
+}
+
+// NewUpdateCommand returns a usable command registered under the parent.
+func NewUpdateCommand(parent common.Registerer, globals *config.Data) *UpdateCommand {
+	var c UpdateCommand
+	c.Globals = globals
+	c.manifest.File.Read(manifest.Filename)
+
+	c.CmdClause = parent.Command("update", "Update a FTP logging endpoint on a Fastly service version")
```
Comment (phamann) [inline pkg/logging/ftp/update.go:43]:
```suggestion
	c.CmdClause = parent.Command("update", "Update an FTP logging endpoint on a Fastly service version")
```

File: pkg/logging/ftp/update.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	c.CmdClause.Flag("name", "The name of the FTP logging object").Short('n').Required().StringVar(&c.Input.Name)
+
+	c.CmdClause.Flag("new-name", "New name of the FTP logging object").Action(c.NewName.Set).StringVar(&c.NewName.Value)
+	c.CmdClause.Flag("address", "An hostname or IPv4 address").Action(c.Address.Set).StringVar(&c.Address.Value)
+	c.CmdClause.Flag("port", "The port number").Action(c.Port.Set).UintVar(&c.Port.Value)
+	c.CmdClause.Flag("user", "The username for the server (can be anonymous)").Action(c.Username.Set).StringVar(&c.Username.Value)
+	c.CmdClause.Flag("password", "The password for the server (for anonymous use an email address)").Action(c.Password.Set).StringVar(&c.Password.Value)
+	c.CmdClause.Flag("path", "The path to upload log files to. If the path ends in / then it is treated as a directory").Action(c.Path.Set).StringVar(&c.Path.Value)
+	c.CmdClause.Flag("period", "How frequently log files are finalized so they can be available for reading (in seconds, default 3600)").Action(c.Period.Set).UintVar(&c.Period.Value)
+	c.CmdClause.Flag("format-version", "@TODO").Action(c.FormatVersion.Set).UintVar(&c.FormatVersion.Value)
```
Comment (phamann) [inline pkg/logging/ftp/update.go:56]: Noting `@TODO` here still need to be done.

**Comment by mccurdyc** [comment]
Force-pushed rebasing master.

**Comment by mccurdyc** [comment]
Force-pushed to create a single commit.

**Comment by mccurdyc** [comment]
Force-pushed to create a single commit.

File: pkg/common/command.go [hunk 6]
```diff
@@ -92,6 +92,12 @@ type OptionalUint struct {
 	Value uint
 }
 
+// OptionalUint models an optional uint8 flag value.
```
Comment (phamann) [inline pkg/common/command.go:95]: Copy/pasta lint error

File: pkg/logging/ftp/create.go (hunk 1 above)
Comment (phamann) [inline pkg/logging/ftp/create.go:26]:
```suggestion
	c.CmdClause = parent.Command("create", "Create an FTP logging endpoint on a Fastly service version").Alias("add")
```

**Comment by mccurdyc** [comment]
Force-pushed condensing into a single commit.

**Comment by mccurdyc** [comment]
Force-pushed rebasing master after merging #62.

**Comment by mccurdyc** [comment]
Force-pushed squashing into a single commit.

**Comment by pteichman** [comment]
Conflict fix (rebase) is at https://github.com/fastly/cli/commit/76b62b18c6909085a0c98c88f98629c13ac3c4b1

**Pull request merged**